Feature :: Canned Tasks

greezybacon opened the pull request:
This implements one of the features outlined in #2826. There it was labelled "Task Templates"
### How to Test
- Checkout the branch from my repo `feature/canned-tasks`, or [visit the branch](https://github.com/greezybacon/osTicket-1.8/tree/feature/canned-tasks) and download from github
- Install or upgrade the system normally
- "Hold on to your butts"

_Please feel free to test and report how you view the software implementing your workflow related challenges._

### Differences from RFC Posting
#### Cancellations

Tasks can be cancelled. This is a special case of PENDING, where the task is closed before it is started. Therefore the PENDING flag is never cleared. This is useful to mark tasks as unnecessary when they are included in a set. A screenshot is included below.

#### Due dates can be based on a related task

Task template due dates can be configured based on the starting date of a related task. For instance, if completing two tasks must happen within a certain period, the second task can use the start time of the first task a reference. A screenshot is included below.

#### Task Template Groups can be windowed

Task template groups can be associated with a department and/or a help topic. This will help workflows for various departments from be accidentally used or cluttering up the UI for unrelated tickets and departments.

### Glossary

I apologize of the verbiage in the pull request differ from the definitions here. That's most of my intention in posting the definitions.

**Task Template** is a skeleton of a single item in a workflow. This might also be called a _Canned Task_
**Task Template Group** is a workflow which includes one or more _Task Templates_ which describe the work items and the interdependency of those items.
**Task Set** a group of tasks which was added to a _Ticket_
**Dependency** is a task which must be completed before a certain task can be started. If task A is a _Dependency_ of task B, then task A must be completed before task B can start. Also, task B depends on task A
**Dependent** is a task on which a task depends. That is, if task B depends on task A, then task B is a _Dependent_ of task A 

### Outstanding
- [x] Installation is missing initial ticket
- [x] New ticket `Issue Summary` does not stick
- [x] Implement enabling and disabling of task templates
- [x] Implement enabling and disabling of task template groups
- [x] Add manual _start_. If a task can be started, then allow an agent to manually start the task
- [x] Polish and test the upgrader from `develop-next`
- [ ] New tasks in a set are not sorted to add to the bottom--especially if there have been deletions
- [x] Delete active task in template group seems not to work
- [x] Add support to close a task and cancel all children / descendents (ie. nothing to do)
- [ ] After adding a template to a workflow, have option to "add another"
- [ ] In task template listing, add "select all", '... none', '... toggle'
- [ ] Add "Canned Responses" to task UI
- [x] Dropdown in ticket thread for [add task set] should not show if no sets are enabled

### Screenshot Tour
#### Add a new task template group.

<img width="950" alt="screen shot 2016-02-02 at 11 26 40 am" src="https://cloud.githubusercontent.com/assets/672074/12762906/2e7ff724-c9b7-11e5-9de8-63f92554c09d.png">
Visit _Admin Panel > Manage > Canned Tasks_ and add from the popup dialog. Task template groups are logical groups of tasks for a particular work flow, such as installing a new piece of equipment.

#### Add a template to the new group

<img width="839" alt="screen shot 2016-02-02 at 11 28 29 am" src="https://cloud.githubusercontent.com/assets/672074/12762959/7e37f88e-c9b7-11e5-9555-d026dc7b03dc.png">

Use the `Add Template` button to visit a new page to add a template to the group.

<img width="662" alt="screen shot 2016-02-02 at 11 28 45 am" src="https://cloud.githubusercontent.com/assets/672074/12762989/9dd26044-c9b7-11e5-81ab-f910f24cf27b.png">

The page looks very simple to the ad-hoc add-task dialog. A few extra fields are added including a complex due date field, dependency selection, and additional forms

#### Floating due dates

Building off the _Pending_ status, the due dates of templated tasks can be configured to have automated due dates based on several references, including other tasks.

<img width="940" alt="screen shot 2016-02-02 at 1 32 04 pm" src="https://cloud.githubusercontent.com/assets/672074/12763183/a8df55b8-c9b8-11e5-9ff9-e44154630bcc.png">

#### Tasks in the template group can be manually sorted and are grouped based on dependencies

<img width="952" alt="screen shot 2016-02-02 at 1 23 58 pm" src="https://cloud.githubusercontent.com/assets/672074/12763023/d1c97810-c9b7-11e5-9d45-092dd0e216d0.png">

#### Add a task template group (workflow) to a ticket

The `Add Task` button on the ticket view page is outfitted with a new drop-down, which allows adding tasks as a group to a ticket.

<img width="960" alt="screen shot 2016-02-02 at 1 34 58 pm" src="https://cloud.githubusercontent.com/assets/672074/12763040/ec0d62c2-c9b7-11e5-87b8-659053c70b4c.png">

The dialog allows for a selection of the template group, which only shows the workflows configured for the help topic and/or department of the ticket currently.

<img width="631" alt="screen shot 2016-02-02 at 1 35 49 pm" src="https://cloud.githubusercontent.com/assets/672074/12763061/0d27376c-c9b8-11e5-890c-a926b7e96564.png">

Additionally, an existing task can be selected as an _Initial Dependency_. In such a case the first tasks in the workflow are not started immediately. Instead, the are delayed until the referenced task is completed.

#### Tasks attached to tickets are grouped

<img width="951" alt="screen shot 2016-02-02 at 2 04 03 pm" src="https://cloud.githubusercontent.com/assets/672074/12763118/543f251a-c9b8-11e5-8c15-351c53658783.png">

Also, this is the first mention of a new task state, _Pending_, which indicates that a task is created, but not yet started. This also implies a deferred due date.

#### Task in a set show a table of all related tasks

<img width="949" alt="screen shot 2016-02-02 at 1 44 10 pm" src="https://cloud.githubusercontent.com/assets/672074/12763259/04b44cfe-c9b9-11e5-904f-479079a0badb.png">

This table is omitted from the view when rendering in the ticket-view page, because the listing there already breaks tasks into groups.

#### Pending tasks start automatically

When a task is closed which is part of a set, all tasks which now have all their dependencies satisfied are automatically started immediately. If they have floating due dates, the due date is calculated based on the automated start time.

<img width="964" alt="screen shot 2016-02-02 at 1 45 32 pm" src="https://cloud.githubusercontent.com/assets/672074/12763281/244a663e-c9b9-11e5-8fd6-d749a18b4651.png">

Notice how the _Created_ time may differ from the _Started_ time 

<img width="794" alt="screen shot 2016-02-02 at 1 46 53 pm" src="https://cloud.githubusercontent.com/assets/672074/12763320/5aaead2a-c9b9-11e5-9646-bcd5c862b0e4.png">

#### Task dependencies are highlighted

At the top of the task view page, a listing of a tasks dependencies are shown along with links to the other tasks and some visual information on their status.

<img width="828" alt="screen shot 2016-02-02 at 1 55 48 pm" src="https://cloud.githubusercontent.com/assets/672074/12763325/69d1bf04-c9b9-11e5-8d63-dc587f7fb938.png">

#### Cancelling tasks

As shown above, tasks which have not yet been started can be cancelled. Cancelling is analogous to closing; except that the task is never started. This is useful if the task does not apply to the modeled workflow. For instance, if installing this particular equipment does not replace any other equipment, then decommissioning will likely not be necessary.

Commit:
tasks: Start with the initial concept of canned tasks

Commit:
thread: Add `started` event

This is used to differentiate between when a task is created and when the
task is started. It will help explain the timeline gaps for dependent canned
tasks, where several tasks are created at the same time but are started in
some hierarchical order rather than simultaneously.

It also adds database support for an `other` event type, which may help in
adding new event types for extensions without the need for special database
modifications.

Commit:
task: Add cancel support for pending tasks

Commit:
tasks: Fix ticket with missing summary, add manual start

Commit:
search: Fix nasty criteria layout

balojs commented:
+1

robintoy commented:
+1

robintoy commented:
This looks amazing!!!!
Two thumbs very high up on this!

nathanzeng commented:
![Uploading image.png…]()
Hi @greezybacon. My apologies but i couldn't create any new tasks or task groups after the initial installation (i was stuck in the loading page). Any guidance? Many thanks.

cluck commented:
I can confirm that "Add Task Template Set" is not working.

Log file contains: **PHP Fatal error:  Call to undefined method VerySimpleModel::create() in /var/www/lib/osticket/include/class.task.php on line 2531**

Looks like the code requires a function `VerySimpleModel::create()` which does not exist.

cluck commented:
By analogy to how the code in class.page.php looks like, changing calls from `parent::create()` to `new static()` seems to make this work.

```
diff --git a/include/class.task.php b/include/class.task.php
index f80a248..c035597 100644
--- a/include/class.task.php
+++ b/include/class.task.php
@@ -1437,7 +1437,7 @@ class Task extends TaskModel implements RestrictedAccess, Threadable {
     static function create($vars=false) {
         global $cfg;

-        $task = parent::create(($vars ?: array()) + array(
+        $task = new static(($vars ?: array()) + array(
             'flags' => self::ISPENDING | self::ISOPEN,
             'number' => $cfg->getNewTaskNumber(),
             'created' => new SqlFunction('NOW'),
@@ -2213,7 +2213,7 @@ class TaskTemplate extends VerySimpleModel {
     }

     static function create($vars=false) {
-        $tpl = parent::create($vars);
+        $tpl = new static($vars);
         $tpl->created = SqlFunction::NOW();
         return $tpl;
     }
@@ -2528,7 +2528,7 @@ class TaskTemplateGroup extends VerySimpleModel {
     }

     static function create($vars=false) {
-        $inst = parent::create($vars);
+        $inst = new static($vars);
         $inst->created = SqlFunction::NOW();
         return $inst;
     }
@@ -2674,7 +2674,7 @@ class TaskSet extends VerySimpleModel {
     }

     static function create($vars=false) {
-        $set = parent::create($vars);
+        $set = new static($vars);
         $set->created = SqlFunction::NOW();
         return $set;
     }
@@ -2800,7 +2800,7 @@ class TaskThread extends ObjectThread {
         assert($task !== false);

         $id = is_object($task) ? $task->getId() : $task;
-        $thread = parent::create(array(
+        $thread = new static(array(
                     'object_id' => $id,
                     'object_type' => ObjectModel::OBJECT_TYPE_TASK
                     ));
```

Commit:
Fix compatibility with upstream changes

Commit:
orm: Cleanup expression API

Everything should extend from SqlExpression, and two expression combined
with an operator should be a BinaryExpression.

Commit:
tasks: Implement enable, disable, and delete for templates

greezybacon commented:
All right gentlemen, I made an effort to reconcile with the upstream changes.

Commit:
tasks: Implement enable, disable, delete for template groups

Chefkeks commented:
@greezybacon
Great to see some progress here :)
I really wish I had time to test this and other interesting/promising PRs (like custom queues) :(
Anyway, keep it going! :+1:

Cheers!
Michael

cluck commented:
Very grateful for the work! I'm looking at this patch with interest to understand the latest coding standards for osTicket and how to extend the code :) I'm also looking forward to contribute for linking tasks to a catalog of system operations (shell scripts running in background or AMQP messages sent out, while responses are filled back into the ticket asynchronously completing or failing the task).

For now I found some more issues:

Deleting a "Task Template" on the "Task Template Set" edit page effectively deletes the task template, but sends me back to the "Task Template Sets" page instead of refreshing the current "Task Template Set" page.

The "Add Template" button on the "Task Template Set" does an Ajax call but nothing happens on the UI. No error log on the server, a 200 OK response a large `<script type="text/javascript">` in the body.

Editing a "Help Topics" under "New ticket options", the "Add Canned Tasks" remains empty. The previously created Task Template Sets with unrestricted visibility are not shown.

Probably a "Filter Action" should be offered to add canned tasks to tickets.

I'm also thinking how (canned) tasks can be linked to system scripts running outside the web application.

greezybacon commented:
> Editing a "Help Topics" under "New ticket options", the "Add Canned Tasks" remains empty. The previously created Task Template Sets with unrestricted visibility are not shown

Have you enabled your templates?

> Probably a "Filter Action" should be offered to add canned tasks to tickets.

Great idea!

Commit:
tasks: Add filter action to add canned tasks to new tickets

Commit:
tasks: lint: Fix PHP7 compatibility issues

Commit:
task: Fix task create

greezybacon commented:
![image](https://cloud.githubusercontent.com/assets/672074/18106534/ca1c6ea8-6ec9-11e6-8320-89995bf68e99.png)

And I fixed the [Add Template] button. @cluck thanks very much for your testing!

cluck commented:
I've done another round of testing and things look promising.

There is at least one more `::create()` that I've hit in include/class.task.php, around line 2213; changing it to a `new TaskTemplateForm()` worked for me. Note that grep'ing through class.task.php there are other `::create()` that probably need be changed.

When deleting a Task Template from a Set, the dialog says that deleted data can not be recovered. This is probably the osTicket standard message but is misleading, as the tasks instantiated from the templates are probably (need to check) preserving their data and it's just the template that is deleted.

After confirming the deletion of a Task Template from a Set I get **Warning: call_user_func_array() expects parameter 1 to be a valid callback, cannot access protected method TaskTemplate::setFlag() in /var/www/lib/osticket/include/class.orm.php on line 722**, and while the message says **Successfully deleted one task template**, the task template is not deleted.

The custom forms on tasks are neither shown nor available to be edited still? This would be likely the most welcome of all features about tasks, because tasks tend to be repetitive and thus require structured data (more than tickets actually).

Commit:
tasks: Fix save of forms on taks templates

greezybacon commented:
@cluck thanks for the continued feedback. I think I was able to remedy the issues with saving and displaying the forms on the task templates. I was not able to reproduce the delete issue. I've rebase and re-pushed several times... Could that be part of the problem?

TL;DR: Be careful, `create` is still valid for many things in the osTicket codebase ...

_I imagine the PHP developers think it's a necessary feature..._  Starting with PHP7, it is a bug if a subclass redefines a method whose call signature is not compatible with the same method defined in any parent. For instance:

```php
class A {
    function say_hello() {
        print "hi";
    }
}

class B extends A {
    function say_hello(array $options) {
        if (!$options['silent'])
            print "hey there";
    }
}
```
```
 $ php php7-example.php
PHP Warning:  Declaration of B::say_hello(array $options) should be compatible with A::say_hello() in /path/to/php7-example.php on line 13
```
This became a problem for us. The ORM defined a `create` which received an array of items. Those items would become the fields saved to the database. However, something like a `Ticket` had a different set of parameters. This makes sense from a coding standpoint, because creating a Ticket is different from creating a User which is also different from creating Syslog entry. So the signature was different. When porting to PHP7 this became an issue. So the creation functionality was moved to the constructor in 1ba40e35d67bb36d71cc292707858c3f211668ba.

This allows each class to have its constructor match that of the ORM's constructor; however, each class can have a specific signature for its `create` method which is specific for creating that kind of object.

So, many of the classes still utilize a `create` method for legacy reasons. However, most of the new stuff has been migrated to perform new-object setup in the object's constructor.

_As a sidenote, and to make things even more confusing, when fetching data from the database, the class's constructor is bypassed to prevent confusion between new and existing objects from the database._

greezybacon commented:
I've rebased this on the v1.10 release for those who are still working on testing

Commit:
tasks: Add upgrade script for canned tasks

nilschrishahn commented:
It looks very nice, but how I can install it into a working 1.10 installation? I tried to upgrade but it didn't work.
Many thanks,
Chris

balojs commented:
thanks a lot @greezybacon

dexmox commented:
@greezybacon - Great efforts.

Can you please clarify the build say 1.8-git but this looks identical to 1.10 which I am running at the moment. 
3 posts back you said you rebased it on v1.10 release is this just theme or is this core?

Look forward to your response;
Have a great day!

Regs,
JV.

greezybacon commented:
@dexmox, when we branched to osTicket v1.8, we decided not to branch again. So the branch tag in the source code still reads `1.8-git`. However, there is also a MAJOR_VERSION tag added now in `bootstrap.php` which should be 1.10. So don't be confused. I synchonized it with the `develop` branch, so it's a feature on top of v1.10.

dexmox commented:
I seem to be getting a hang when deleting tickets and cancelling tasks, some errors show in the php.log I am not able to get it but I will post up later can I confirm this is not an issue for you ?

motux commented:
Thanks a lot, great job

motux commented:
I try this function but the tab of Visibility and Assignment of the Task, when i save the changes, the system don't save anything.
The table ost_task_template_form continue to be empty,.

rtrice81 commented:
Is there anyway to merge with with v1.10? and how installed this version and it is working great I am just wondering when it will be merged with v1.10 or added to the core install of osTicket?

Also I have noticed that the canned tasks will not save the assignee or the days until the task is due

Commit:
forms: Fix missing data for inline form fields

This fixes an issue where the due date for a canned task could not be updated.

greezybacon commented:
@rtrice81 @motux thanks for the feedback. I was able to resolve the issues saving the due date and assignee information.

rtrice81 commented:
@greezybacon tested and confirm that your fix now works as it should for us.
Something else I noticed was that there is no image beside Canned Tasks in the admin panel see screenshot here https://i.imgur.com/B1C2fk5.png
Thank you for this awesome feature, hopefully it will get put into the next stable release

F3000 commented:
@greezybacon 

Is this feteaure included in 1.10 ? 

I can't find this feature. 

Is it possible , when i assign a ticket to a specific help topic , osticket generates automatically a task ?

ntozier commented:
Q: Is this feature included in 1.10 ?
A: No.  It is also not included in 1.10.1.  This is an open feature that has not been merged with the main branch at this time.

F3000 commented:
ok thanks . So i have to wait

greezybacon commented:
@F3000 Yes, that feature is in this patch. Navigate to Admin Panel -> Manage -> Help Topics. Then select the relevant help topic. Then select the group of canned tasks which should be generated with new tickets for such a help topic.

davidegiunchidiennea commented:
I've upgraded to ost v1.10.1 and applied this patch: it works good and it's very useful, thank you  @greezybacon !
This patch is available since feb 2016, has been upgraded to the ost versions of this years (it's mantained), it colmate an ost gap that the user's are asking for since a lot of years (search github, ost forum), the ost's mantainer on 2015 said that this is in the todo list ( https://github.com/osTicket/osTicket/issues/2093#issuecomment-103681978 ), now the question is:

Why this patch has not been included in the ost's core?
There's a roadmap to include it in the next ost's version?

regards

ntozier commented:
@davidegiunchidiennea the osTicket roadmap is not public, and I doubt that you will get an answer to those questions.

davidegiunchidiennea commented:
Ok, there isn't any roadmap.
There is any *idea* to include this patch in any next version?
or it will never be included/it's difficult that will ever be included?

JediKev commented:
@davidegiunchidiennea 

>There is any idea to include this patch in any next version?

Possibly.

>or it will never be included/it's difficult that it will ever be included?

It will be included but we don't have a set version or date of release. Stay tuned!

davidegiunchidiennea commented:
@JediKev that's good, Thank you!

TuxNDK commented:
I have a new install 1.10.1 and add this patch - not it shows 1.8-git. I can't create tasks as canned or single task as normal. I just return to dialog for create new ticket when I try.
Im new on osticket and maybe I patch the worn way.
I just download repo and overwrite my installation with this. Then it ask for DB upgrade, as I did without problem.

Rujinraj commented:
I am using the latest version v1.10.1. How can i patch in my current system with this canned task feature.

Kindly help

greezybacon commented:
@Rukinraj, I have rebased this branch on the upstream develop branch (v1.10.2). You should be able to check it out or deploy it and run the upgrader. Although, you should *definitely* try it in a test environment with a copy of your data first, do user acceptance testing, and then consider upgrading your production system.

greezybacon commented:
Aight party people. I've rebased this on 1.11.x in hopes of continuing the discussion of getting this to become a major feature. I've tested a new install thus far. Absolutely no guarantees on upgrading from 1.11!

wizard210 commented:
@greezybacon this would be a game changer for schools I believe. I know it would for my department. We are looking for a way to set up a process for approving or denying apps for our devices. 

Do you think this would be okay in a production environment yet?

greezybacon committed:
tasks: Hide "Add Task Group"

Hide the "Attach Task Template Group" drop down on the [Tasks] tab of
the thread view if there are no task groups available for ticket.

greezybacon committed:
tasks: Create set of tasks in dependency order

Sort the tasks being created when a template group is attached to a
ticket in the order requested by the tasks while also considering the
declared dependencies. This will help the order of the tasks make more
sense when viewed in the task listing on the ticket thread page.

greezybacon committed:
tasks: Show related tasks in hierarchical view

In addition to sorting the related tasks based on their declared order
and dependencies, indent the dependencies in the way it's viewed in the
admin panel when editing the template group.

greezybacon committed:
tasks: Implement cancellation of a task

greezybacon committed:
tasks: Add feature to cascade cancellation

So, when a task is cancelled, offer the option to also cancel all
dependent tasks, and all tasks dependent on those tasks. The list is
presented in the dialog as a checkbox list, and, by default, all tasks
are selected to be cancelled.

Cancelling only some dependent tasks currently has the caveat that tasks
which depend on a cancelled task might not be started when they are able
to be started. In such a case, an agent can manually start or cancel the
task at a later date.

greezybacon commented:
For those watch this feature, I've rebased it on the 1.12.x branch, so it should work with the latest features and security fixes in osTicket. I've also fixed a couple sorting issues and added the ability to cascade-cancel tasks (also cancel all subordinate tasks).

@wizard210 I would generally recommend making a copy of your production database and upgrading it for use with this feature. Any comments and feedback are welcome!

edgarnadal commented:
Is this awesome PR ever going  to be merged? @greezybacon

greezybacon commented:
@edgarnadal thanks of the ping! It ultimately isn't up to me, though. Have you tested the canned tasks feature yet in this pull request? Do you have any comments or suggestions?

justdave commented:
I ran across a need for this (several tasks associated with onboarding new employees - same 9 tasks on every ticket of this type).  But I'm already running osTicket 1.14.  Any chance of this getting updated for 1.14.x?  According to the automerge box the existing patch has conflicts.

apinner84 commented:
Hi @greezybacon this looks absolutely perfect for what I'm trying to do. Any chance of a rebase onto the latest release?

Can't believe this feature hasn't been merged already!

greezybacon committed:
Properly add `assigned` event on staff assignment

greezybacon commented:
@apinner84, @JasonBee- I've rebased it on the develop branch. Any and all comments are welcome.

nachtbraker commented:
Will this ever be added to the main line?
